forcedotcom/source-deploy-retrieve
# fix: offer suggestions for unresolved metadata types

### What does this PR do?
In the past, a misspelled metadata file would fall back to `EmailServicesFunction` because that is the default for the `.xml` suffix. This change will prevent that incorrect resolved type and also offer suggestions to try similar metadata types. 

### Step 1 - Validation

* Create and bootstrap [Dreamhouse](https://github.com/trailheadapps/dreamhouse-lwc)
* Run `sfdx project generate manifest --source-dir force-app`
    * Note that the `package.xml` has the following `Layout` entry
```
<types>
    <members>Broker__c-Broker Layout</members>
    <members>Property__c-Property Layout</members>
    <name>Layout</name>
</types>
```

### Step 2 - Reproduce bug

* Rename the `Property` layout to have a capital `L` on `.Layout-meta.xml`
    * Run `mv force-app/main/default/layouts/Property__c-Property\ Layout.layout-meta.xml force-app/main/default/layouts/Property__c-Property\ Layout.Layout-meta.xml`
* Run `sfdx project generate manifest --source-dir force-app`
* Note that the generated manifest
    * Completes "successfully"
    * The `Label` type no longer contains `Property__c-Property\ Layout` 
    * There is an incorrect entry for `EmailServicesFunction`
```
<types>
    <members>Property__c-Property Layout</members>
    <name>EmailServicesFunction</name>
</types>
```
* This is because [this lookup](https://github.com/forcedotcom/source-deploy-retrieve/blob/11218d7e51baadaa01ecc4aba5a96c213b00d20a/src/resolve/metadataResolver.ts#L204) is using the `.xml` filename suffix which was claimed by `EmailServicesFunction` [here](https://github.com/forcedotcom/source-deploy-retrieve/blob/11218d7e51baadaa01ecc4aba5a96c213b00d20a/src/registry/metadataRegistry.json#L3681)


### Step 3 - Testing Changes

* Pull this branch
* Run `yarn link`
* `cd` to `plugin-deploy-retrieve`
* Run `yarn link @salesforce/source-deploy-retrieve`
* Run `sfdx plugins link`
* Run `sfdx plugins` to confirm plugin is linked
* Run `sfdx project generate manifest --source-dir force-app`
* Confirm that you see a message like:

<img width="1258" alt="Screenshot 2023-04-28 at 10 46 44 AM" src="https://user-images.githubusercontent.com/1715111/235244509-139706c4-761c-45d9-ab4e-4e3ccda3089f.png">

* Note that this is not specific to case. You can misspell a metadata type and it will try to find the closest match.
    * Try renaming the layout to `lyout-meta.xml`



### What issues does this PR fix or reference?
[@W-12731813@](https://gus.my.salesforce.com/apex/ADM_WorkLocator?bugorworknumber=W-12731813)
also fixes [@W-12484289@](https://gus.lightning.force.com/a07EE00001KZ8xkYAD)
